nixos/modules/services/continuous-integration/gitea-actions-runner.nix

```
@@ -0,0 +1,176 @@
+    # explicit container usage
+    (any (label: hasInfix label "docker://") instance.labels)
+  ) cfg.instances;
+
+  usesDocker = usesContainers && config.virtualisation.docker.enable;
+  usesPodman = usesContainers && config.virtualisation.podman.enable;
+in
+{
+  meta.maintainers = with lib.maintainers; [
+    hexa
+  ];
+
+  options.services.gitea-actions-runner = with types; {
+    package = mkPackageOptionMD pkgs "gitea-actions-runner" { };
+
+    instances = mkOption {
+      default = {};
+      description = lib.mdDoc ''
+        Gitea Actions Runner instances.
+      '';
+      type = attrsOf (submodule {
+        options = {
+          enable = mkEnableOption (lib.mdDoc "Gitea Actions Runner instance");
+
+          name = mkOption {
+            type = str;
+            example = literalExpression "config.networking.hostName";
+            description = lib.mdDoc ''
+              The name identifying the runner instance towards the Gitea/Forgejo instance.
+            '';
+          };
+
+          url = mkOption {
+            type = str;
+            example = "https://forge.example.com";
+            description = lib.mdDoc ''
+              Base URL of your Gitea/Forgejo instance.
+            '';
+          };
+
+          token = mkOption {
+            type = nullOr str;
+            default = null;
+            description = lib.mdDoc ''
+              Plain token to register at the configured Gitea/Forgejo instance.
+            '';
+          };
+
+          tokenFile = mkOption {
+            type = nullOr path;
+            default = null;
+            description = lib.mdDoc ''
+              Path to an environment file, containing the `TOKEN` environment
+              variable, that holds a token to register at the configured
+              Gitea/Forgejo instance.
+            '';
+          };
+
+          labels = mkOption {
+            type = listOf str;
+            example = literalExpression ''
+              [
+                "host:host"
+                "nixos:docker://nixos/nix:latest"
+                "ubuntu-rolling:docker://ubuntu:rolling"
+                "ubuntu-lts:docker//ubuntu:latest"
+              ]
+            '';
+            description = lib.mdDoc ''
+              Labels used to map jobs to their runtime environment.
+            '';
+            apply = concatStringsSep ",";
+          };
+        };
+      });
+    };
+  };
+
+  config = mkIf (cfg.instances != {}) {
+    assertions = let
+      tokenXorTokenFile = instance:
+        (instance.token == null && instance.tokenFile != null) ||
+        (instance.token != null && instance.tokenFile == null);
+    in  [ {
+      assertion = any tokenXorTokenFile (attrValues cfg.instances);
+      message = "Instances of gitea-actions-runner can have `token` or `tokenFile`, not both.";
+    } ];
+
+    systemd.services = let
+      mkRunnerService = name: instance:
+        nameValuePair "gitea-runner-${escapeSystemdPath name}" {
```

We should assert this as well IMHO

```suggestion
      mkRunnerService = name: instance:
        let
          usesContainers = hasInfix "docker://" instance.labels;
          usesDocker = usesContainers && config.virtualisation.docker.enable;
          usesPodman = usesContainers && config.virtualisation.podman.enable;
        in
        nameValuePair "gitea-runner-${escapeSystemdPath name}" {
```

---

Assert what exactly?

---

Oops, sorry.

I think it would make sense to actually assert `config.virtualisation.docker.enable` and `config.virtualisation.podman.enable` when `usesContainers = true`.

Currently, an instance using the default labels (either empty, or from my suggestion) is unable to start and crashes with
```
Error: cannot ping the docker daemon, does it running? Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

Asserting `config.virtualisation.docker.enable || config.virtualisation.podman.enable` when `usesContainers = true` would move this error from runtime to eval.

Hope this makes a bit more sense now :)